pipelines.md

```diff
@@ -0,0 +1,358 @@
+* Tagging is not expressive enough
+* Any non-trivial tagging is complex and unintuitive (i.e., the learning curve is not smooth).
+
+This proposal will not fix all of the above, but will address some issues. In particular, I plan to think in more depth about the more complex tagging requirements.
+
+# Proposed changes
+
+## Core syntax and semantics
+
+The pipeline operator becomes a method call operator.
+
+Alternative: we could replace `|>` with `.` - this would be closer to most other languages, but I feel would lose some of KCL's character.
+
+### Functions may have an explicit receiver
+
+In the declaration, this is the first argument of a function and must use a distinguished keyword (most PLs use `self` or `this`) or symbol (we could reuse `#` either alone or as a decorator on a name, e.g., `#sketch`, or any other sigil). I prefer a keyword, I have no preference which. Note that in other PLs the receiver is used for dispatch of a method, not just the syntax of calls. That is not proposed yet.
+
+In a call, the lhs of the `|>` operator becomes the receiver. If the function is called without the pipeline, then the argument must be specified.
+
+E.g.,
+
+```
+f = fn (this) => { ... }
+
+startSketchOn(...) |> f()
+f(startSketchOn(...))
+```
+
+In both calls, the result of `startSketchOn` is passed to `f` as `this`.
+
+
+### Remove `#`
+
+The `#` operator is no longer supported.
+
+In the common case, the use of `#` is replaced with use of a receiver (see above).
+
+The other use cases for `#` I can find are for tagging (e.g., `profileStartX(%)`) and nested geometry (e.g., `hole(circle(..., #), #)``). I'm optimistic that by refining the design of std and implementing some extensions to tagging, we can reduce the need for these uses (both are future work). In the meantime, these uses can be implemented by breaking the pipeline with an intermediate variable, e.g.,
+
+```
+// Current KCL
+sketch006 = startSketchOn('XZ')
+  |> startProfileAt([0.1, 1], %)
+  |> line([0.1, 0], %)
+  |> angledLineToX({ angle: 10, to: 0.05 }, %)
+  |> yLine(10, %)
+  |> line([0.6, 0], %)
+  |> yLine(-.05, %)
+  |> tangentialArc({ radius: 0.6, offset: -90 }, %)
+  |> lineTo([profileStartX(%), profileStartY(%)], %)
+  |> close(%)
+  |> revolve({ axis: 'y' }, %)
+
+// Proposed KCL
+sketch006 = startSketchOn('XZ')
+  |> startProfileAt([0.1, 1])
+  |> line([0.1, 0])
+  |> angledLineToX({ angle: 10, to: 0.05 })
+  |> yLine(10)
+  |> line([0.6, 0])
+  |> yLine(-.05)
+  |> tangentialArc({ radius: 0.6, offset: -90 })
+sketch007 = sketch006
+  |> lineTo([profileStartX(sketch006), profileStartY(sketch006)])
+  |> close()
+  |> revolve({ axis: 'y' })
+```
+
+See below ('Tagging') for a further improvement to this example.
+
+#### Alternative
+
+Retain `#` or replace it with a keyword (which I would prefer) for the rare cases where it is required (c.f., the current situation where it is frequently required).
```

Comment:
I'd prefer not to, if it's not needed often. It's one more thing to learn and one more thing for beginners to see and get confused by. I believe any remaining requirement can be met by factoring out a variable, possibly using `as` to avoid splitting the pipeline (and using a variable is self-documenting in the way that `%` is not).